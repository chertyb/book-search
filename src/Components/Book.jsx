import React from 'react';
import BookShape, { defaultBookShape } from '../Shape/BookShape';

const Book = (props) => {
  const { data } = props;

  return (
    <div className="book">
      <img src={data.thumbnail} alt="thumbnail" />
      <p title="Title">
        {'Title: '}
        {data.title}
      </p>
      <p title="Authors">
        {'Authors: '}
        {data.authors}
      </p>
      <p title="Categories">
        {'Categories: '}
        {data.categories}
      </p>
      <p title="Description">
        {'Description: '}
        {data.description}
      </p>
    </div>
  );
};

Book.propTypes = {
  data: BookShape,
};

Book.defaultProps = {
  data: defaultBookShape,
};

export default Book;
