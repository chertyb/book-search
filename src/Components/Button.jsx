import React from 'react';
import PropTypes from 'prop-types';

const Button = (props) => {
  const { text, handleClick } = props;
  return <button type="button" onClick={handleClick}>{text}</button>;
};

Button.propTypes = {
  text: PropTypes.string,
  handleClick: PropTypes.func,
};

Button.defaultProps = {
  text: '',
  handleClick: () => {},
};

export default Button;
