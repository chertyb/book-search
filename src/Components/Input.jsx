import React from 'react';
import PropTypes from 'prop-types';

const Input = (props) => {
  const { value, handleInputChange } = props;
  return <input type="text" value={value} onChange={handleInputChange} />;
};

Input.propTypes = {
  value: PropTypes.string,
  handleInputChange: PropTypes.func,
};

Input.defaultProps = {
  value: '',
  handleInputChange: () => {},
};

export default Input;
