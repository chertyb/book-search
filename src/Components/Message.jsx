import React from 'react';
import PropTypes from 'prop-types';

const Message = (props) => {
  const { message, type } = props;
  return <p className={type}>{message}</p>;
};

Message.propTypes = {
  message: PropTypes.string,
  type: PropTypes.string,
};

Message.defaultProps = {
  message: '',
  type: '',
};

export default Message;
