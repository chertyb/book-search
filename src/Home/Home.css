#book-display {
  width: 80%;
  height: 500px;
  border: 1px solid black;
  margin: auto;
  margin-top: 30px;
  overflow-y: auto;
}
