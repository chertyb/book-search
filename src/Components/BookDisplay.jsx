import React from 'react';
import PropTypes from 'prop-types';
import Book from './Book';
import BookShape from '../Shape/BookShape';

const BookDisplay = (props) => {
  const { books, totalBooks, loadMore } = props;

  const handleScroll = (e) => {
    const divBottom = e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight;
    if (divBottom && books.length < totalBooks) {
      loadMore(false);
    }
  };

  return (
    <div id="book-display" onScroll={handleScroll}>
      {books
      && books.map((book) => <Book data={book} key={`book-${book.id}`} />)}
    </div>
  );
};

BookDisplay.propTypes = {
  books: PropTypes.arrayOf(BookShape),
  loadMore: PropTypes.func,
  totalBooks: PropTypes.number,
};

BookDisplay.defaultProps = {
  books: [],
  loadMore: () => {},
  totalBooks: 0,
};

export default BookDisplay;
