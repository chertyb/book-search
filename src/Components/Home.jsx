import React from 'react';
import Input from './Input';
import Button from './Button';
import Message from './Message';
import BookDisplay from './BookDisplay';
import fetchBooks, { maxResults, removeDuplicates, formatBooksfromAPI } from '../Utils/Books';

class Home extends React.Component {
  constructor() {
    super();
    this.state = {
      searchInput: '',
      books: [],
      isLoading: false,
      error: null,
      currentIndex: 0,
      totalBooks: 0,
    };
    this.handleSearchInput = this.handleSearchInput.bind(this);
    this.handleSearchButton = this.handleSearchButton.bind(this);
    this.searchBooks = this.searchBooks.bind(this);
  }

  async searchBooks(isNewSearch) {
    const { currentIndex, books, searchInput } = this.state;
    if (searchInput === '') {
      this.setState({ error: 'Please enter a query', books: [] });
    } else {
      this.setState({ isLoading: true, error: null });
      // if new search, empty display so index 0 and empty books
      const newIndex = isNewSearch ? 0 : currentIndex;
      const currentBooks = newIndex ? books : [];

      fetchBooks(searchInput, currentIndex)
        .then((result) => {
          const totalBooks = result.totalItems ? result.totalItems : 0;
          const nextBooks = result.items ? formatBooksfromAPI(result.items) : [];
          // Google Book API does not always return same list with same request
          // so we remove duplicates
          const filteredBooks = removeDuplicates([...currentBooks, ...nextBooks]);
          this.setState({
            books: filteredBooks,
            isLoading: false,
            currentIndex: currentIndex + maxResults,
            totalBooks,
          });
        })
        .catch((error) => this.setState({ error: error.toString(), isLoading: false }));
    }
  }

  handleSearchInput(e) {
    this.setState({ searchInput: e.target.value });
  }

  handleSearchButton() {
    this.searchBooks(true);
  }

  render() {
    const {
      searchInput, isLoading, error, books, totalBooks,
    } = this.state;
    return (
      <div>
        <h1>Book Search App</h1>
        <Input value={searchInput} handleInputChange={this.handleSearchInput} />
        <Button text="Search" handleClick={this.handleSearchButton} />
        {isLoading
          && <Message message="Loading..." type="loading" />}
        {error
          && <Message message={error} type="error" />}
        <BookDisplay books={books} loadMore={this.searchBooks} totalBooks={totalBooks} />
      </div>
    );
  }
}

export default Home;
