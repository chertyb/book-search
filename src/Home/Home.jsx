import React from 'react';
import './Home.css';
import InputComponent from '../Components/InputComponent';
import ButtonComponent from '../Components/ButtonComponent';
import MessageComponent from '../Components/MessageComponent';
import BookDisplayComponent from '../Components/BookDisplayComponent';
import fetchBooks, { maxResults, formatBooksfromAPI } from '../Utils/Books';

class Home extends React.Component {
  constructor() {
    super();
    this.state = {
      searchInput: '',
      books: [],
      isLoading: false,
      error: null,
      currentIndex: 0,
      totalBooks: 0,
    };
    this.handleSearchInput = this.handleSearchInput.bind(this);
    this.handleSearchButton = this.handleSearchButton.bind(this);
    this.searchBooks = this.searchBooks.bind(this);
  }

  searchBooks(isNewSearch) {
    const { currentIndex, books, searchInput } = this.state;
    if (!searchInput) {
      this.setState({ error: 'Please enter a query', books: [] });
    } else {
      this.setState({ isLoading: true, error: null });
      // if new search, empty display so index 0 and empty books
      const newIndex = isNewSearch ? 0 : currentIndex;
      const currentBooks = newIndex ? books : [];

      fetchBooks(searchInput, currentIndex)
        .then((result) => {
          const totalBooks = result.totalItems ? result.totalItems : 0;
          const nextBooks = result.items ? formatBooksfromAPI(result.items) : [];
          this.setState({
            books: [...currentBooks, ...nextBooks],
            isLoading: false,
            currentIndex: currentIndex + maxResults,
            totalBooks,
          });
        })
        .catch((error) => this.setState({ error: error.toString(), isLoading: false }));
    }
  }

  handleSearchInput(e) {
    this.setState({ searchInput: e.target.value });
  }

  handleSearchButton() {
    this.searchBooks(true);
  }

  render() {
    const {
      searchInput, isLoading, error, books, totalBooks,
    } = this.state;
    return (
      <div>
        <h1>Book Search App</h1>
        <InputComponent value={searchInput} handleInputChange={this.handleSearchInput} />
        <ButtonComponent text="Search" handleClick={this.handleSearchButton} />
        {isLoading
          && <MessageComponent message="Loading..." />}
        {error
          && <MessageComponent message={error} />}
        <BookDisplayComponent books={books} loadMore={this.searchBooks} totalBooks={totalBooks} />
      </div>
    );
  }
}

export default Home;
