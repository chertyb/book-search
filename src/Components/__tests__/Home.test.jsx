import React from 'react';
import ReactDom from 'react-dom';
import { mount } from '../../enzyme';
import Home from '../Home';
import Input from '../Input';
import Button from '../Button';
import Message from '../Message';
import BookDisplay from '../BookDisplay';
import Book from '../Book';
import { defaultBookShape } from '../../Shape/BookShape';

describe('Home', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Home />);
  });
  afterEach(() => {
    wrapper.unmount();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<Home />, div);
    expect(div).toMatchSnapshot();
    ReactDom.unmountComponentAtNode(div);
  });

  it('should display empty Input component', () => {
    const input = wrapper.find(Input);
    expect(input.length).toEqual(1);
    expect(input.prop('value')).toEqual('');
  });

  it('should change state after input value change', () => {
    const newValue = 'Harry Potter';
    const input = wrapper.find(Input);
    input.simulate('change', { target: { value: newValue } });
    expect(wrapper.state().searchInput).toEqual(newValue);
  });

  it('should display Button component', () => {
    expect(wrapper.find(Button).length).toEqual(1);
  });

  it('should display empty BookDisplay component', () => {
    const bookDisplay = wrapper.find(BookDisplay);
    expect(bookDisplay.length).toEqual(1);
    expect(bookDisplay.length).toEqual(1);
  });

  it('should display 0 books', () => {
    expect(wrapper.find(Book).exists()).toEqual(false);
  });

  it('should display error message when search button clicked with no input', () => {
    wrapper.setState({ searchInput: '' });
    wrapper.find(Button).simulate('click');
    expect(wrapper.find(Message).length).toEqual(1);
    expect(wrapper.find(Message).prop('type')).toEqual('error');
  });

  it('should display loading message when search button clicked with input', () => {
    wrapper.setState({ searchInput: 'Harry Potter' });
    wrapper.find(Button).simulate('click');
    expect(wrapper.find(Message).length).toEqual(1);
    expect(wrapper.find(Message).prop('type')).toEqual('loading');
  });

  it('should display fetched books', () => {
    wrapper.setState({ books: [{ ...defaultBookShape, id: 'id1' }, { ...defaultBookShape, id: 'id2' }] });
    expect(wrapper.find(Book).length).toEqual(2);
  });
});
