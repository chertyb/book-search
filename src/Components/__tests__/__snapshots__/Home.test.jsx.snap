// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders without crashing 1`] = `
<div>
  <div>
    <h1>
      Book Search App
    </h1>
    <input
      type="text"
      value=""
    />
    <button
      type="button"
    >
      Search
    </button>
    <div
      id="book-display"
    />
  </div>
</div>
`;
